web3/middleware/filter.py

```diff
@@ -377,3 +383,281 @@ def middleware(method: RPCEndpoint, params: Any) -> RPCResponse:
             return make_request(method, params)
 
     return middleware
+
+
+# --- async --- #
+
+
+async def async_iter_latest_block(
+    w3: "Web3", to_block: Optional[Union[BlockNumber, LatestBlockParam]] = None
+) -> AsyncIterable[BlockNumber]:
+    """Returns a generator that dispenses the latest block, if
+    any new blocks have been mined since last iteration.
+
+    If there are no new blocks or the latest block is greater than
+    the ``to_block`` None is returned.
+
+    >>> new_blocks = iter_latest_block(w3, 0, 10)
+    >>> next(new_blocks)  # Latest block = 0
+    0
+    >>> next(new_blocks)  # No new blocks
+    >>> next(new_blocks)  # Latest block = 1
+    1
+    >>> next(new_blocks)  # Latest block = 10
+    10
+    >>> next(new_blocks)  # latest block > to block
+    """
+    _last = None
+
+    is_bounded_range = to_block is not None and to_block != "latest"
+
+    while True:
+        latest_block = await w3.eth.block_number  # type: ignore
+        # type ignored b/c is_bounded_range prevents unsupported comparison
+        if is_bounded_range and latest_block > to_block:
+            yield None
+        #  No new blocks since last iteration.
+        if _last is not None and _last == latest_block:
+            yield None
+        else:
+            yield latest_block
+        _last = latest_block
+
+
+async def async_iter_latest_block_ranges(
+    w3: "Web3",
+    from_block: BlockNumber,
+    to_block: Optional[Union[BlockNumber, LatestBlockParam]] = None,
+) -> AsyncIterable[Tuple[Optional[BlockNumber], Optional[BlockNumber]]]:
+    """Returns an iterator unloading ranges of available blocks
+
+    starting from `fromBlock` to the latest mined block,
+    until reaching toBlock. e.g.:
+
+
+    >>> blocks_to_filter = iter_latest_block_ranges(w3, 0, 50)
+    >>> next(blocks_to_filter)  # latest block number = 11
+    (0, 11)
+    >>> next(blocks_to_filter)  # latest block number = 45
+    (12, 45)
+    >>> next(blocks_to_filter)  # latest block number = 50
+    (46, 50)
+    """
+    latest_block_iterator = async_iter_latest_block(w3, to_block)
+    async for latest_block in latest_block_iterator:
+        if latest_block is None:
+            yield (None, None)
+        elif from_block > latest_block:
+            yield (None, None)
+        else:
+            yield (from_block, latest_block)
+            from_block = BlockNumber(latest_block + 1)
+
+
+async def async_get_logs_multipart(
+    w3: "Web3",
+    startBlock: BlockNumber,
```

Comment:
We should use `start_block` and `stop_block` here, as well as change these in the synchronous `get_logs_multipart`. This can be in a separate PR if preferred but I think it's a pretty small change that would be simple enough to review if it gets its own commit.

Reply:
:heavy_check_mark: